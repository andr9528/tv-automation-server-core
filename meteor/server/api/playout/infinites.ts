import * as _ from 'underscore'
import { DBPart, Part } from '../../../lib/collections/Parts'
import { Piece, Pieces } from '../../../lib/collections/Pieces'
import { asyncCollectionFindFetch } from '../../../lib/lib'
import { PartInstance, PartInstanceId } from '../../../lib/collections/PartInstances'
import { PieceInstance } from '../../../lib/collections/PieceInstances'
import { RundownPlaylist } from '../../../lib/collections/RundownPlaylists'
import { getAllOrderedPartsFromCache, selectNextPart, getSelectedPartInstancesFromCache } from './lib'
import { CacheForRundownPlaylist } from '../../DatabaseCaches'
import { saveIntoCache } from '../../DatabaseCache'
import {
	getPieceInstancesForPart as libgetPieceInstancesForPart,
	getPlayheadTrackingInfinitesForPart as libgetPlayheadTrackingInfinitesForPart,
	buildPiecesStartingInThisPartQuery,
	buildPastInfinitePiecesForThisPartQuery,
} from '../../../lib/rundown/infinites'
import { profiler } from '../profiler'

// /** When we crop a piece, set the piece as "it has definitely ended" this far into the future. */
export const DEFINITELY_ENDED_FUTURE_DURATION = 1 * 1000

/**
 * We can only continue adlib onEnd infinites if we go forwards in the rundown. Any distance backwards will clear them.
 * */
function canContinueAdlibOnEndInfinites(
	playlist: RundownPlaylist,
	orderedParts: Part[],
	previousPartInstance: PartInstance | undefined,
	part: DBPart
): boolean {
	if (previousPartInstance && playlist) {
		const span = profiler.startSpan('canContinueAdlibOnEndInfinites')
		// TODO - if we don't have an index for previousPartInstance, what should we do?

		const expectedNextPart = selectNextPart(playlist, previousPartInstance, orderedParts)
		if (expectedNextPart) {
			if (expectedNextPart.part._id === part._id) {
				// Next part is what we expect, so take it
				return true
			} else {
				const partIndex = orderedParts.findIndex((p) => p._id === part._id)
				if (partIndex >= expectedNextPart.index) {
					if (span) span.end()
					// Somewhere after the auto-next part, so we can use that
					return true
				} else {
					if (span) span.end()
					// It isnt ahead, so we cant take it
					return false
				}
			}
		} else {
			if (span) span.end()
			// selectNextPart gave nothing, so we must be at the end?
			return false
		}
	} else {
		// There won't be anything to continue anyway..
		return false
	}
}

function getIdsBeforeThisPart(cache: CacheForRundownPlaylist, nextPart: DBPart) {
	const span = profiler.startSpan('getIdsBeforeThisPart')
	// Note: This makes the assumption that nextPart is a part found in this cache
	const partsBeforeThisInSegment = cache.Parts.findFetch({
		segmentId: nextPart.segmentId,
		_rank: { $lt: nextPart._rank },
	}).map((p) => p._id)
	const currentSegment = cache.Segments.findOne(nextPart.segmentId)
	const segmentsBeforeThisInRundown = currentSegment
		? cache.Segments.findFetch({
				rundownId: nextPart.rundownId,
				_rank: { $lt: currentSegment._rank },
		  }).map((p) => p._id)
		: []

	if (span) span.end()
	return {
		partsBeforeThisInSegment,
		segmentsBeforeThisInRundown,
	}
}

export async function fetchPiecesThatMayBeActiveForPart(
	cache: CacheForRundownPlaylist,
	part: DBPart
): Promise<Piece[]> {
	const span = profiler.startSpan('fetchPiecesThatMayBeActiveForPart')

	const thisPiecesQuery = buildPiecesStartingInThisPartQuery(part)
	const pPiecesStartingInPart = cache.Pieces.initialized
		? Promise.resolve(cache.Pieces.findFetch(thisPiecesQuery))
		: asyncCollectionFindFetch(Pieces, thisPiecesQuery)

	const { partsBeforeThisInSegment, segmentsBeforeThisInRundown } = getIdsBeforeThisPart(cache, part)

	const infinitePiecesQuery = buildPastInfinitePiecesForThisPartQuery(
		part,
		partsBeforeThisInSegment,
		segmentsBeforeThisInRundown
	)
	const pInfinitePieces = cache.Pieces.initialized
		? Promise.resolve(cache.Pieces.findFetch(infinitePiecesQuery))
		: asyncCollectionFindFetch(Pieces, infinitePiecesQuery)

	const [piecesStartingInPart, infinitePieces] = await Promise.all([pPiecesStartingInPart, pInfinitePieces])
	if (span) span.end()
	return [...piecesStartingInPart, ...infinitePieces]
}

export function syncPlayheadInfinitesForNextPartInstance(
	cache: CacheForRundownPlaylist,
	playlist: RundownPlaylist
): void {
	const span = profiler.startSpan('syncPlayheadInfinitesForNextPartInstance')
	const { nextPartInstance, currentPartInstance } = getSelectedPartInstancesFromCache(cache, playlist)
	if (nextPartInstance && currentPartInstance) {
		const infinites = getPlayheadTrackingInfinitesForPart(cache, playlist, currentPartInstance, nextPartInstance)

		saveIntoCache(
			cache.PieceInstances,
			{
				partInstanceId: nextPartInstance._id,
				'infinite.fromPreviousPlayhead': true,
			},
			infinites
		)
	}
	if (span) span.end()
}

function getPlayheadTrackingInfinitesForPart(
	cache: CacheForRundownPlaylist,
	playlist: RundownPlaylist,
	playingPartInstance: PartInstance,
	nextPartInstance: PartInstance
): PieceInstance[] {
	const span = profiler.startSpan('getPlayheadTrackingInfinitesForPart')
	const { partsBeforeThisInSegment, segmentsBeforeThisInRundown } = getIdsBeforeThisPart(cache, nextPartInstance.part)

	const orderedParts = getAllOrderedPartsFromCache(cache, playlist)

	const canContinueAdlibOnEnds = canContinueAdlibOnEndInfinites(
		playlist,
		orderedParts,
		playingPartInstance,
		nextPartInstance.part
	)
	const playingPieceInstances = cache.PieceInstances.findFetch((p) => p.partInstanceId === playingPartInstance._id)

	const res = libgetPlayheadTrackingInfinitesForPart(
		new Set(partsBeforeThisInSegment),
		new Set(segmentsBeforeThisInRundown),
		playingPartInstance,
		playingPieceInstances,
		nextPartInstance.part,
		nextPartInstance._id,
		canContinueAdlibOnEnds,
		false
	)
	if (span) span.end()
	return res
}

export function getPieceInstancesForPart(
	cache: CacheForRundownPlaylist,
	playlist: RundownPlaylist,
	playingPartInstance: PartInstance | undefined,
	part: DBPart,
	possiblePieces: Piece[],
	newInstanceId: PartInstanceId,
	isTemporary: boolean
): PieceInstance[] {
	const span = profiler.startSpan('getPieceInstancesForPart')
	const { partsBeforeThisInSegment, segmentsBeforeThisInRundown } = getIdsBeforeThisPart(cache, part)

	const orderedParts = getAllOrderedPartsFromCache(cache, playlist)
	const playingPieceInstances = playingPartInstance
		? cache.PieceInstances.findFetch((p) => p.partInstanceId === playingPartInstance._id)
		: []

	const canContinueAdlibOnEnds = canContinueAdlibOnEndInfinites(playlist, orderedParts, playingPartInstance, part)

	const res = libgetPieceInstancesForPart(
		playingPartInstance,
		playingPieceInstances,
		part,
		new Set(partsBeforeThisInSegment),
		new Set(segmentsBeforeThisInRundown),
		possiblePieces,
		orderedParts.map((part) => part._id),
		newInstanceId,
		canContinueAdlibOnEnds,
		isTemporary
	)
	if (span) span.end()
	return res
}
