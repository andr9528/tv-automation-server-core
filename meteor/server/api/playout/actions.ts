import { Meteor } from 'meteor/meteor'
import * as _ from 'underscore'
import { logger } from '../../logging'
import { Rundown, Rundowns, RundownHoldState } from '../../../lib/collections/Rundowns'
import { Parts } from '../../../lib/collections/Parts'
import { Studio, StudioId, Studios } from '../../../lib/collections/Studios'
import { PeripheralDevices, PeripheralDevice } from '../../../lib/collections/PeripheralDevices'
import { PeripheralDeviceAPI } from '../../../lib/api/peripheralDevice'
import { getCurrentTime, getRandomId, waitForPromise } from '../../../lib/lib'
import { loadShowStyleBlueprint } from '../blueprints/cache'
import { RundownContext } from '../blueprints/context'
import {
	setNextPart,
	onPartHasStoppedPlaying,
	selectNextPart,
	getSelectedPartInstancesFromCache,
	getStudioFromCache,
	getAllOrderedPartsFromCache,
} from './lib'
import { updateTimeline } from './timeline'
import { IngestActions } from '../ingest/actions'
import { getActiveRundownPlaylistsInStudio } from './studio'
import { RundownPlaylists, RundownPlaylist } from '../../../lib/collections/RundownPlaylists'
import { PartInstances } from '../../../lib/collections/PartInstances'
import { CacheForRundownPlaylist } from '../../DatabaseCaches'
import { profiler } from '../profiler'

export function activateRundownPlaylist(
	cache: CacheForRundownPlaylist,
	rundownPlaylist: RundownPlaylist,
	rehearsal: boolean
): void {
	logger.info('Activating rundown ' + rundownPlaylist._id + (rehearsal ? ' (Rehearsal)' : ''))

	rehearsal = !!rehearsal

	let newRundown = cache.RundownPlaylists.findOne(rundownPlaylist._id) // fetch new from db, to make sure its up to date

	if (!newRundown) throw new Meteor.Error(404, `Rundown "${rundownPlaylist._id}" not found!`)
	rundownPlaylist = newRundown

	let studio = cache.activationCache.getStudio()

	const anyOtherActiveRundowns = getActiveRundownPlaylistsInStudio(cache, studio._id, rundownPlaylist._id)

	if (anyOtherActiveRundowns.length) {
		// logger.warn('Only one rundown can be active at the same time. Active rundowns: ' + _.map(anyOtherActiveRundowns, rundown => rundown._id))
		throw new Meteor.Error(
			409,
			'Only one rundown can be active at the same time. Active rundown playlists: ' +
				_.map(anyOtherActiveRundowns, (playlist) => playlist._id),
			JSON.stringify(_.map(anyOtherActiveRundowns, (playlist) => playlist._id))
		)
	}

	cache.RundownPlaylists.update(rundownPlaylist._id, {
		$set: {
			active: true,
			rehearsal: rehearsal,
			activeInstanceId: getRandomId(),
		},
	})

	// Re-Initialize the ActivationCache now when the rundownPlaylist is active
	const rundownsInPlaylist = cache.Rundowns.findFetch()
	waitForPromise(cache.activationCache.initialize(rundownPlaylist, rundownsInPlaylist))

	let rundown: Rundown | undefined

	if (!rundownPlaylist.nextPartInstanceId) {
		const firstPart = selectNextPart(rundownPlaylist, null, getAllOrderedPartsFromCache(cache, rundownPlaylist))
		setNextPart(cache, rundownPlaylist, firstPart ? firstPart.part : null)
	} else {
		const nextPartInstance = cache.PartInstances.findOne(rundownPlaylist.nextPartInstanceId)
		if (!nextPartInstance)
			throw new Meteor.Error(404, `Could not find nextPartInstance "${rundownPlaylist.nextPartInstanceId}"`)
		rundown = cache.Rundowns.findOne(nextPartInstance.rundownId)
		if (!rundown) throw new Meteor.Error(404, `Could not find rundown "${nextPartInstance.rundownId}"`)
	}

	updateTimeline(cache, studio._id)

	cache.defer((cache) => {
		if (!rundown) return // if the proper rundown hasn't been found, there's little point doing anything else
		const { blueprint } = loadShowStyleBlueprint(waitForPromise(cache.activationCache.getShowStyleBase(rundown)))
		const context = new RundownContext(rundown, cache, undefined)
		context.wipeCache()
		if (blueprint.onRundownActivate) {
			Promise.resolve(blueprint.onRundownActivate(context)).catch(logger.error)
		}
	})
}
export function deactivateRundownPlaylist(cache: CacheForRundownPlaylist, rundownPlaylist: RundownPlaylist): void {
	const rundown = deactivateRundownPlaylistInner(cache, rundownPlaylist)

	updateTimeline(cache, rundownPlaylist.studioId)

	cache.defer((cache) => {
		if (rundown) {
			const { blueprint } = loadShowStyleBlueprint(
				waitForPromise(cache.activationCache.getShowStyleBase(rundown))
			)
			if (blueprint.onRundownDeActivate) {
				Promise.resolve(blueprint.onRundownDeActivate(new RundownContext(rundown, cache, undefined))).catch(
					logger.error
				)
			}
		}
	})
}
export function deactivateRundownPlaylistInner(
	cache: CacheForRundownPlaylist,
	rundownPlaylist: RundownPlaylist
): Rundown | undefined {
	const span = profiler.startSpan('deactivateRundownPlaylistInner')
	logger.info(`Deactivating rundown playlist "${rundownPlaylist._id}"`)

	const { currentPartInstance, nextPartInstance } = getSelectedPartInstancesFromCache(cache, rundownPlaylist)

	let rundown: Rundown | undefined
	if (currentPartInstance) {
		// defer so that an error won't prevent deactivate
		Meteor.setTimeout(() => {
			rundown = Rundowns.findOne(currentPartInstance.rundownId)

			if (rundown) {
				IngestActions.notifyCurrentPlayingPart(rundown, null)
			} else {
				logger.error(
					`Could not find owner Rundown "${currentPartInstance.rundownId}" of PartInstance "${currentPartInstance._id}"`
				)
			}
		}, 40)
	} else {
		if (nextPartInstance) {
			rundown = cache.Rundowns.findOne(nextPartInstance.rundownId)
		}
	}

	if (currentPartInstance) onPartHasStoppedPlaying(cache, currentPartInstance, getCurrentTime())

	cache.RundownPlaylists.update(rundownPlaylist._id, {
		$set: {
			active: false,
			previousPartInstanceId: null,
			currentPartInstanceId: null,
			holdState: RundownHoldState.NONE,
		},
	})
	// rundownPlaylist.currentPartInstanceId = null
	// rundownPlaylist.previousPartInstanceId = null
	setNextPart(cache, rundownPlaylist, null)

	if (currentPartInstance) {
		cache.PartInstances.update(currentPartInstance._id, {
			$set: {
				'timings.takeOut': getCurrentTime(),
			},
		})
	}
	if (span) span.end()
	return rundown
}
/**
 * Prepares studio before a broadcast is about to start
 * @param studio
 * @param okToDestoryStuff true if we're not ON AIR, things might flicker on the output
 */
export function prepareStudioForBroadcast(
	okToDestoryStuff: boolean,
	rundownPlaylistToBeActivated: RundownPlaylist
): void {
	if (!rundownPlaylistToBeActivated.studioId)
		throw new Meteor.Error(500, `Playlist "${rundownPlaylistToBeActivated._id}" has no studioId!`)
	logger.info('prepareStudioForBroadcast ' + rundownPlaylistToBeActivated.studioId)

	let playoutDevices = PeripheralDevices.find({
		studioId: rundownPlaylistToBeActivated.studioId,
		type: PeripheralDeviceAPI.DeviceType.PLAYOUT,
	}).fetch()

	_.each(playoutDevices, (device: PeripheralDevice) => {
		PeripheralDeviceAPI.executeFunction(
			device._id,
			(err) => {
				if (err) {
					logger.error(err)
				} else {
					logger.info('devicesMakeReady OK')
				}
			},
			'devicesMakeReady',
			okToDestoryStuff,
			rundownPlaylistToBeActivated._id
		)
	})
}
/**
 * Makes a studio "stand down" after a broadcast
 * @param studio
 * @param okToDestoryStuff true if we're not ON AIR, things might flicker on the output
 */
export function standDownStudio(cache: CacheForRundownPlaylist, studio: Studio, okToDestoryStuff: boolean): void {
	logger.info('standDownStudio ' + studio._id)

	let playoutDevices = waitForPromise(cache.activationCache.getPeripheralDevices()).filter(
		(d) => d.type === PeripheralDeviceAPI.DeviceType.PLAYOUT
	)

	_.each(playoutDevices, (device: PeripheralDevice) => {
		PeripheralDeviceAPI.executeFunction(
			device._id,
			(err) => {
				if (err) {
					logger.error(err)
				} else {
					logger.info('devicesStandDown OK')
				}
			},
			'devicesStandDown',
			okToDestoryStuff
		)
	})
}
