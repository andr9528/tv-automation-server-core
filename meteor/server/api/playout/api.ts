import { Meteor } from 'meteor/meteor'
import { registerClassToMeteorMethods } from '../../methods'
import { NewPlayoutAPI, PlayoutAPIMethods } from '../../../lib/api/playout'
import { ServerPlayoutAPI } from './playout'
import { getCurrentTime, makePromise } from '../../../lib/lib'
import { logger } from '../../logging'
import { RundownPlaylistId } from '../../../lib/collections/RundownPlaylists'
import { PartInstanceId } from '../../../lib/collections/PartInstances'
import { PartId } from '../../../lib/collections/Parts'
import { PieceId } from '../../../lib/collections/Pieces'
import { StudioId } from '../../../lib/collections/Studios'
import { PieceInstanceId } from '../../../lib/collections/PieceInstances'
import { ClientAPI } from '../../../lib/api/client'
import { MethodContextAPI } from '../../../lib/api/methods'
import { Settings } from '../../../lib/Settings'
import { SegmentId } from '../../../lib/collections/Segments'

class ServerPlayoutAPIClass extends MethodContextAPI implements NewPlayoutAPI {
	rundownPrepareForBroadcast(playlistId: RundownPlaylistId) {
		return makePromise(() => ServerPlayoutAPI.prepareRundownPlaylistForBroadcast(this, playlistId))
	}
	rundownResetRundown(playlistId: RundownPlaylistId) {
		return makePromise(() => ServerPlayoutAPI.resetRundownPlaylist(this, playlistId))
	}
	rundownResetAndActivate(playlistId: RundownPlaylistId, rehearsal?: boolean) {
		return makePromise(() => ServerPlayoutAPI.resetAndActivateRundownPlaylist(this, playlistId, rehearsal))
	}
	rundownActivate(playlistId: RundownPlaylistId, rehearsal: boolean) {
		return makePromise(() => ServerPlayoutAPI.activateRundownPlaylist(this, playlistId, rehearsal))
	}
	rundownDeactivate(playlistId: RundownPlaylistId) {
		return makePromise(() => ServerPlayoutAPI.deactivateRundownPlaylist(this, playlistId))
	}
	reloadRundownPlaylistData(playlistId: RundownPlaylistId) {
		return makePromise(() => ServerPlayoutAPI.reloadRundownPlaylistData(this, playlistId))
	}
	pieceTakeNow(
		playlistId: RundownPlaylistId,
		partInstanceId: PartInstanceId,
		pieceInstanceIdOrPieceIdToCopy: PieceInstanceId | PieceId
	) {
		return makePromise(() =>
			ServerPlayoutAPI.pieceTakeNow(this, playlistId, partInstanceId, pieceInstanceIdOrPieceIdToCopy)
		)
	}
	rundownTake(playlistId: RundownPlaylistId) {
		return makePromise(() => ServerPlayoutAPI.takeNextPart(this, playlistId))
	}
	rundownSetNext(playlistId: RundownPlaylistId, partId: PartId, timeOffset?: number | undefined) {
		return makePromise(() => ServerPlayoutAPI.setNextPart(this, playlistId, partId, true, timeOffset))
	}
	rundownSetNextSegment(playlistId: RundownPlaylistId, segmentId: SegmentId | null) {
		return makePromise(() => ServerPlayoutAPI.setNextSegment(this, playlistId, segmentId))
	}
	rundownMoveNext(playlistId: RundownPlaylistId, horisontalDelta: number, verticalDelta: number) {
		return makePromise(() => ServerPlayoutAPI.moveNextPart(this, playlistId, horisontalDelta, verticalDelta, true))
	}
	rundownActivateHold(playlistId: RundownPlaylistId) {
		return makePromise(() => ServerPlayoutAPI.activateHold(this, playlistId))
	}
	rundownDisableNextPiece(rundownPlaylistId: RundownPlaylistId, undo?: boolean) {
		return makePromise(() => ServerPlayoutAPI.disableNextPiece(this, rundownPlaylistId, undo))
	}
	segmentAdLibPieceStart(
		rundownPlaylistId: RundownPlaylistId,
		partInstanceId: PartInstanceId,
		pieceId: PieceId,
		queue: boolean
	) {
		return makePromise(() =>
			ServerPlayoutAPI.segmentAdLibPieceStart(this, rundownPlaylistId, partInstanceId, pieceId, queue)
		)
	}
	rundownBaselineAdLibPieceStart(
		rundownPlaylistId: RundownPlaylistId,
		partInstanceId: PartInstanceId,
		pieceId: PieceId,
		queue: boolean
	) {
		return makePromise(() =>
			ServerPlayoutAPI.rundownBaselineAdLibPieceStart(this, rundownPlaylistId, partInstanceId, pieceId, queue)
		)
	}
	sourceLayerOnPartStop(
		rundownPlaylistId: RundownPlaylistId,
		partInstanceId: PartInstanceId,
		sourceLayerIds: string[]
	) {
		return makePromise(() =>
			ServerPlayoutAPI.sourceLayerOnPartStop(this, rundownPlaylistId, partInstanceId, sourceLayerIds)
		)
	}
	sourceLayerStickyPieceStart(playlistId: RundownPlaylistId, sourceLayerId: string) {
		return makePromise(() => ServerPlayoutAPI.sourceLayerStickyPieceStart(this, playlistId, sourceLayerId))
	}
	updateStudioBaseline(studioId: StudioId) {
		return makePromise(() => ServerPlayoutAPI.updateStudioBaseline(this, studioId))
	}
	shouldUpdateStudioBaseline(studioId: StudioId) {
		return makePromise(() => ServerPlayoutAPI.shouldUpdateStudioBaseline(this, studioId))
	}
	switchRouteSet(studioId: StudioId, routeSetId: string, state: boolean) {
		return makePromise(() => ServerPlayoutAPI.switchRouteSet(this, studioId, routeSetId, state))
	}
}
registerClassToMeteorMethods(PlayoutAPIMethods, ServerPlayoutAPIClass, false)

if (!Settings.enableUserAccounts) {
	// Temporary methods
	Meteor.methods({
		debug__printTime: () => {
			let now = getCurrentTime()
			logger.debug(new Date(now))
			return now
		},
	})
}
