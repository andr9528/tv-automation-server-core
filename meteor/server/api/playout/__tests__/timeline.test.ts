import '../../../../__mocks__/_extendJest'
import { testInFiber } from '../../../../__mocks__/helpers/jest'
import { fixSnapshot } from '../../../../__mocks__/helpers/snapshot'
import {
	setupDefaultStudioEnvironment,
	DefaultEnvironment,
	setupDefaultRundownPlaylist,
	setupMockPeripheralDevice,
} from '../../../../__mocks__/helpers/database'
import { Rundowns, Rundown } from '../../../../lib/collections/Rundowns'
import '../api'
import { Timeline } from '../../../../lib/collections/Timeline'
import { ServerPlayoutAPI } from '../playout'
import { updateTimeline } from '../timeline'
import { RundownPlaylists, RundownPlaylist } from '../../../../lib/collections/RundownPlaylists'
import { protectString, waitForPromise } from '../../../../lib/lib'
import { MethodContext } from '../../../../lib/api/methods'
import { initCacheForNoRundownPlaylist, initCacheForRundownPlaylistFromStudio } from '../../../DatabaseCaches'
import { PeripheralDeviceAPI } from '../../../../lib/api/peripheralDevice'

const DEFAULT_CONTEXT: MethodContext = {
	userId: null,
	isSimulation: false,
	connection: {
		id: 'mockConnectionId',
		close: () => {},
		onClose: () => {},
		clientAddress: '127.0.0.1',
		httpHeaders: {},
	},
	setUserId: () => {},
	unblock: () => {},
}

describe('Timeline', () => {
	let env: DefaultEnvironment
	beforeEach(() => {
		env = setupDefaultStudioEnvironment()
		setupMockPeripheralDevice(
			PeripheralDeviceAPI.DeviceCategory.PLAYOUT,
			PeripheralDeviceAPI.DeviceType.PLAYOUT,
			PeripheralDeviceAPI.SUBTYPE_PROCESS,
			env.studio
		)
	})
	testInFiber('non-existing studio', () => {
		expect(() => {
			const studioId = protectString('asdf')
			const cache = waitForPromise(initCacheForNoRundownPlaylist(studioId))
			updateTimeline(cache, protectString('asdf'))
		}).toThrowError(/not found/i)
	})
	testInFiber('Basic rundown', () => {
		const { rundownId: rundownId0, playlistId: playlistId0 } = setupDefaultRundownPlaylist(env)
		expect(rundownId0).toBeTruthy()
		expect(playlistId0).toBeTruthy()

		const getRundown0 = () => {
			return Rundowns.findOne(rundownId0) as Rundown
		}
		const getPlaylist0 = () => {
			return RundownPlaylists.findOne(playlistId0) as RundownPlaylist
		}
		expect(getRundown0()).toBeTruthy()
		expect(getPlaylist0()).toBeTruthy()

		const parts = getRundown0().getParts()

		expect(getPlaylist0()).toMatchObject({
			active: false,
			rehearsal: false,
		})

		{
			// Prepare and activate in rehersal:
			ServerPlayoutAPI.activateRundownPlaylist(DEFAULT_CONTEXT, playlistId0, false)
			const { currentPartInstance, nextPartInstance } = getPlaylist0().getSelectedPartInstances()
			expect(currentPartInstance).toBeFalsy()
			expect(nextPartInstance).toBeTruthy()
			expect(nextPartInstance!.part._id).toEqual(parts[0]._id)
			expect(getPlaylist0()).toMatchObject({
				active: true,
				rehearsal: false,
				currentPartInstanceId: null,
				// nextPartInstanceId: parts[0]._id,
			})
		}

		{
			// Take the first Part:
			ServerPlayoutAPI.takeNextPart(DEFAULT_CONTEXT, playlistId0)
			const { currentPartInstance, nextPartInstance } = getPlaylist0().getSelectedPartInstances()
			expect(currentPartInstance).toBeTruthy()
			expect(nextPartInstance).toBeTruthy()
			expect(currentPartInstance!.part._id).toEqual(parts[0]._id)
			expect(nextPartInstance!.part._id).toEqual(parts[1]._id)
			// expect(getPlaylist0()).toMatchObject({
			// 	currentPartInstanceId: parts[0]._id,
			// 	nextPartInstanceId: parts[1]._id,
			// })
		}

		let cache = waitForPromise(initCacheForRundownPlaylistFromStudio(getRundown0().studioId))

		updateTimeline(cache, getRundown0().studioId)
		waitForPromise(cache.saveAllToDatabase())

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()

		cache = waitForPromise(initCacheForRundownPlaylistFromStudio(getRundown0().studioId))

		const currentTime = 100 * 1000
		updateTimeline(cache, getRundown0().studioId, currentTime)
		waitForPromise(cache.saveAllToDatabase())

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()

		{
			// Deactivate rundown:
			ServerPlayoutAPI.deactivateRundownPlaylist(DEFAULT_CONTEXT, playlistId0)
			expect(getPlaylist0()).toMatchObject({
				active: false,
				currentPartInstanceId: null,
				nextPartInstanceId: null,
			})
		}

		expect(fixSnapshot(Timeline.find().fetch())).toMatchSnapshot()
	})
})
