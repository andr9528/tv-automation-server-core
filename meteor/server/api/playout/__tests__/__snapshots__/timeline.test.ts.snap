// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline Basic rundown 1`] = `
Array [
  Object {
    "_id": "mockStudio11",
    "generated": 12345,
    "timeline": Array [
      Object {
        "classes": Array [
          "rundown_active",
        ],
        "content": Object {
          "deviceType": 0,
        },
        "enable": Object {
          "while": 1,
        },
        "id": "playlist_randomId9002_status",
        "layer": "rundown_status",
        "objectType": "rundown",
      },
      Object {
        "children": Array [],
        "content": Object {
          "deviceType": 0,
          "type": "group",
        },
        "enable": Object {
          "start": "now",
        },
        "id": "part_group_randomId9002_part0_0_randomId9007",
        "isGroup": true,
        "isPartGroup": true,
        "layer": "",
        "objectType": "rundown",
        "priority": 5,
      },
      Object {
        "classes": Array [],
        "content": Object {
          "callBack": "partPlaybackStarted",
          "callBackData": Object {
            "partInstanceId": "randomId9002_part0_0_randomId9007",
            "rundownPlaylistId": "playlist_randomId9002",
          },
          "callBackStopped": "partPlaybackStopped",
          "deviceType": 0,
          "type": "callback",
        },
        "enable": Object {
          "start": 0,
        },
        "id": "part_group_firstobject_randomId9002_part0_0_randomId9007",
        "inGroup": "part_group_randomId9002_part0_0_randomId9007",
        "layer": "group_first_object",
        "objectType": "rundown",
      },
      Object {
        "children": Array [],
        "content": Object {
          "deviceType": 0,
          "type": "group",
        },
        "enable": Object {
          "start": 0,
        },
        "id": "piece_group_randomId9002_part0_0_randomId9007_randomId9002_piece001",
        "inGroup": "part_group_randomId9002_part0_0_randomId9007",
        "isGroup": true,
        "layer": "vt0",
        "metaData": Object {
          "pieceId": "randomId9002_part0_0_randomId9007_randomId9002_piece001",
        },
        "objectType": "rundown",
        "priority": 5,
      },
    ],
    "timelineHash": "randomId9012",
  },
]
`;

exports[`Timeline Basic rundown 2`] = `
Array [
  Object {
    "_id": "mockStudio11",
    "generated": 12345,
    "timeline": Array [
      Object {
        "classes": Array [
          "rundown_active",
        ],
        "content": Object {
          "deviceType": 0,
        },
        "enable": Object {
          "while": 1,
        },
        "id": "playlist_randomId9002_status",
        "layer": "rundown_status",
        "objectType": "rundown",
      },
      Object {
        "children": Array [],
        "content": Object {
          "deviceType": 0,
          "type": "group",
        },
        "enable": Object {
          "setFromNow": true,
          "start": 100000,
        },
        "id": "part_group_randomId9002_part0_0_randomId9007",
        "isGroup": true,
        "isPartGroup": true,
        "layer": "",
        "objectType": "rundown",
        "priority": 5,
      },
      Object {
        "classes": Array [],
        "content": Object {
          "callBack": "partPlaybackStarted",
          "callBackData": Object {
            "partInstanceId": "randomId9002_part0_0_randomId9007",
            "rundownPlaylistId": "playlist_randomId9002",
          },
          "callBackStopped": "partPlaybackStopped",
          "deviceType": 0,
          "type": "callback",
        },
        "enable": Object {
          "start": 0,
        },
        "id": "part_group_firstobject_randomId9002_part0_0_randomId9007",
        "inGroup": "part_group_randomId9002_part0_0_randomId9007",
        "layer": "group_first_object",
        "objectType": "rundown",
      },
      Object {
        "children": Array [],
        "content": Object {
          "deviceType": 0,
          "type": "group",
        },
        "enable": Object {
          "start": 0,
        },
        "id": "piece_group_randomId9002_part0_0_randomId9007_randomId9002_piece001",
        "inGroup": "part_group_randomId9002_part0_0_randomId9007",
        "isGroup": true,
        "layer": "vt0",
        "metaData": Object {
          "pieceId": "randomId9002_part0_0_randomId9007_randomId9002_piece001",
        },
        "objectType": "rundown",
        "priority": 5,
      },
    ],
    "timelineHash": "randomId9013",
  },
]
`;

exports[`Timeline Basic rundown 3`] = `
Array [
  Object {
    "_id": "mockStudio11",
    "generated": 12345,
    "timeline": Array [
      Object {
        "content": Object {
          "deviceType": 0,
        },
        "enable": Object {
          "start": 0,
        },
        "id": "baseline_version",
        "layer": "baseline_version",
        "metaData": Object {
          "versions": Object {
            "blueprintId": "mockBlueprint7",
            "blueprintVersion": "0.0.0",
            "core": "0.0.0-test",
            "studio": "asdf",
          },
        },
        "objectType": "rundown",
      },
    ],
    "timelineHash": "randomId9017",
  },
]
`;
